src/org/mozilla/fxtest/ServiceBook.groovy

```diff
@@ -0,0 +1,167 @@
+
+    connection.setRequestMethod(verb);
+    connection.setRequestProperty("Content-Type", "application/json");
+    connection.doOutput = true;
+
+    def writer = new OutputStreamWriter(connection.outputStream);
+    writer.write(json);
+    writer.flush();
+    writer.close();
+
+    connection.connect();
+
+    //parse the response
+    HttpResponse resp = new HttpResponse(connection);
+
+    if(resp.isFailure()){
+        error("\n$verb to URL: $requestUrl\n    JSON: $json\n    HTTP Status: $resp.statusCode\n    Message: $resp.message\n    Response Body: $resp.body");
+    }
+
+    return resp;
+}
+
+class HttpResponse {
+    String body;
+    String message;
+    Integer statusCode;
+    boolean failure = false;
+
+    public HttpResponse(HttpURLConnection connection){
+        this.statusCode = connection.responseCode;
+        this.message = connection.responseMessage;
+
+        if(statusCode == 200 || statusCode == 201){
+            this.body = connection.content.text;//this would fail the pipeline if there was a 400
+        }else{
+            this.failure = true;
+            this.body = connection.getErrorStream().text;
+        }
+
+        connection = null; //set connection to null for good measure, since we are done with it
+    }
+}
+
+
+// Servicebook iterator to get operational tests for a given project
+Object getProjectTests(String name) {
+    resp = doGetHttpRequest("http://servicebook.dev.mozaws.net/api/project").body;
+    def jsonSlurper = new JsonSlurperClassic();
+    def projects = jsonSlurper.parseText(resp);
+
+    for (project in projects.data) {
+        if (project.name == name) {
+            echo "The project " + name + " was found!"
+
+            def jenkin_tests = [];
+                for (test in project.tests) {
+                    if (test.jenkins_pipeline) {
+                        echo "Adding one test " + test.name
+                        jenkin_tests << test;
+                    }
+                }
+            return jenkin_tests;
+        }
+    }
+    echo "The project " + name + " was not found"
+    return null;
+}
+
+
+
+def validURL(url) {
+    allowedOrgs = ['Kinto', 'mozilla', 'mozilla-services', 'tarekziade', 'rpappalax']
+
+    for (allowedOrg in allowedOrgs) {
+        if (url.startsWith('https://github.com/' + allowedOrg + '/')) {
+            return true
+        }
+    }
+    echo url + " is not a valid url"
+    return false
+}
+
+
+def runStage(test) {
+    stage(test.name) {
+        if (validURL(test.url)) {
+            echo "checking out " + test.url + ".git"
+            node {
+                checkout([$class: 'GitSCM',
+                branches: [[name: '*/master']],
+                doGenerateSubmoduleConfigurations: false,
+                extensions: [[$class: 'CleanCheckout']],
+                submoduleCfg: [],
+                userRemoteConfigs: [[url: test.url + '.git']]]
+                )
+            }
+            echo "checked out"
+            node {
```

Comment:
Note that this will run on any executor. If we have a specific agent or label that we wanted to run on, we should do that. I suspect we want a specific docker image.

Reply:
I don't yet know much about these actual tests, and/or where/how they best should run, besides that they should run in Docker, on the slave/execution node (at least in the qa-preprod* new setup).

Can we defer that for now, or should we hold off on merging this in until we have it clarified and codified, here?